File: src/guardrail_hook.ts
/**
 * Anti-Repeat-Errors Skill - Before Tool Call Hook
 *
 * TypeScript hook handler that bridges to the Python guardrail module.
 */

import type { PluginApi, BeforeToolCallContext, HookResult } from "@openclaw/plugin-sdk";
import { spawn } from "child_process";
import * as path from "path";

/**
 * Guardrail result from Python
 */
interface GuardrailResult {
  allowed: boolean;
  modified: boolean;
  tool_name: string;
  original_params: Record<string, any>;
  result_params: Record<string, any>;
  rule_id?: string;
  rule_name?: string;
  action?: string;
  message?: string;
  duration_ms: number;
}

/**
 * Plugin configuration
 */
interface PluginConfig {
  enabled?: boolean;
  guardrailEnabled?: boolean;
  guardrailTimeout?: number;
  guardrailRulesDir?: string;
  rulesDir?: string;
  logLevel?: "debug" | "info" | "warn" | "error";
}

/**
 * Register the before_tool_call hook
 */
export function registerBeforeToolCallHook(api: PluginApi, config: PluginConfig): void {
  const logger = api.logger;

  // Check if guardrail is enabled
  if (config.guardrailEnabled === false) {
    logger.info("Guardrail disabled by config");
    return;
  }

  api.registerHook(
    "before_tool_call",
    async (ctx: BeforeToolCallContext): Promise<HookResult> => {
      const startTime = Date.now();

      try {
        // Build context
        const guardrailContext = {
          session_key: ctx.sessionKey,
          phase: getCurrentPhase(ctx),
          task_type: inferTaskType(ctx),
          message_content: getLastUserMessage(ctx),
        };

        // Call Python guardrail
        const result = await callPythonGuardrail(
          ctx.toolName,
          ctx.toolParams,
          guardrailContext,
          config,
          logger
        );

        // Handle result
        if (!result.allowed) {
          // Blocked
          logger.warn(
            `Tool call blocked by rule ${result.rule_id}: ${result.message}`
          );
          
          // Record metric
          recordGuardrailMetric(result, "blocked", api);
          
          // Return error to agent
          return {
            continue: false,
            error: {
              code: "GUARDRAIL_BLOCKED",
              message: result.message || `Blocked by guardrail rule: ${result.rule_name}`,
              details: {
                rule_id: result.rule_id,
                rule_name: result.rule_name,
              },
            },
          };
        }

        if (result.modified) {
          // Rewritten
          logger.info(
            `Tool call rewritten by rule ${result.rule_id}: ${result.message}`
          );
          
          // Update params
          ctx.toolParams = result.result_params;
          
          // Record metric
          recordGuardrailMetric(result, "rewritten", api);
        }

        if (result.action === "warn" && result.message) {
          // Warning
          logger.warn(`Guardrail warning: ${result.message}`);
          
          // Add warning to context (agent can see it)
          ctx.prependContext = (ctx.prependContext || "") + 
            `\n\n<!-- guardrail-warning -->\n${result.message}\n<!-- /guardrail-warning -->\n`;
          
          recordGuardrailMetric(result, "warned", api);
        }

        if (result.rule_id && result.action === "log") {
          // Logged only
          logger.info(`Guardrail log: rule ${result.rule_id} matched`);
          recordGuardrailMetric(result, "logged", api);
        }

        // Allow to continue
        return { continue: true };

      } catch (error) {
        logger.error(`Guardrail processing failed: ${error}`);
        // Don't block on error
        return { continue: true };
      }
    },
    {
      name: "anti-repeat-errors.before_tool_call",
      description: "Intercepts tool calls and applies guardrail rules",
      priority: 100, // High priority
    }
  );

  logger.info("Registered before_tool_call hook");
}

/**
 * Call Python guardrail CLI
 */
async function callPythonGuardrail(
  toolName: string,
  toolParams: Record<string, any>,
  context: Record<string, any>,
  config: PluginConfig,
  logger: any
): Promise<GuardrailResult> {
  const scriptPath = path.join(__dirname, "guardrail_cli.py");
  const timeout = config.guardrailTimeout || 500; // 500ms default

  // Build args
  const args = [
    scriptPath,
    "--tool-name", toolName,
    "--tool-params", JSON.stringify(toolParams),
    "--context", JSON.stringify(context),
  ];

  // Add rules dir if specified
  const rulesDir = config.guardrailRulesDir || config.rulesDir;
  if (rulesDir) {
    args.push("--rules-dir", rulesDir);
  }

  return new Promise((resolve, reject) => {
    let resolved = false;

    const proc = spawn("python3", args);

    let stdout = "";
    let stderr = "";

    proc.stdout.on("data", (data) => {
      stdout += data.toString();
    });

    proc.stderr.on("data", (data) => {
      stderr += data.toString();
      if (config.logLevel === "debug") {
        logger.debug(`[Python] ${data.toString().trim()}`);
      }
    });

    proc.on("close", (code) => {
      if (resolved) return;
      resolved = true;

      if (code === 0) {
        try {
          const result = JSON.parse(stdout.trim()) as GuardrailResult;
          resolve(result);
        } catch (e) {
          logger.error(`Failed to parse guardrail output: ${stdout}`);
          resolve({
            allowed: true,
            modified: false,
            tool_name: toolName,
            original_params: toolParams,
            result_params: toolParams,
            message: `Parse error: ${e}`,
            duration_ms: 0,
          });
        }
      } else {
        resolve({
          allowed: true, // Don't block on error
          modified: false,
          tool_name: toolName,
          original_params: toolParams,
          result_params: toolParams,
          message: `Guardrail exit code ${code}: ${stderr}`,
          duration_ms: 0,
        });
      }
    });

    proc.on("error", (error) => {
      if (resolved) return;
      resolved = true;
      resolve({
        allowed: true,
        modified: false,
        tool_name: toolName,
        original_params: toolParams,
        result_params: toolParams,
        message: `Spawn error: ${error}`,
        duration_ms: 0,
      });
    });

    // Timeout
    setTimeout(() => {
      if (resolved) return;
      resolved = true;
      proc.kill();
      logger.warn(`Guardrail timeout after ${timeout}ms`);
      resolve({
        allowed: true, // Don't block on timeout
        modified: false,
        tool_name: toolName,
        original_params: toolParams,
        result_params: toolParams,
        message: "Timeout",
        duration_ms: timeout,
      });
    }, timeout);
  });
}

/**
 * Record guardrail metric
 */
function recordGuardrailMetric(
  result: GuardrailResult,
  status: string,
  api: PluginApi
): void {
  if (api.metrics) {
    api.metrics.record("anti_repeat_errors_guardrail", {
      tool_name: result.tool_name,
      rule_id: result.rule_id,
      action: result.action,
      status,
      duration_ms: result.duration_ms,
    });
  }
}

/**
 * Get current phase from context
 */
function getCurrentPhase(ctx: any): number | undefined {
  if (ctx.projectState?.phase) return ctx.projectState.phase;
  const phaseEnv = process.env.ANTI_REPEAT_ERRORS_PHASE;
  if (phaseEnv) {
    const phase = parseInt(phaseEnv, 10);
    if (!isNaN(phase)) return phase;
  }
  return undefined;
}

/**
 * Infer task type from context
 */
function inferTaskType(ctx: any): string | undefined {
  if (ctx.taskType) return ctx.taskType;
  const tools = ctx.sessionHistory?.recentTools || [];
  if (tools.includes("write") || tools.includes("edit")) return "coding";
  if (tools.includes("exec")) return "shell";
  return undefined;
}

/**
 * Get last user message
 */
function getLastUserMessage(ctx: any): string | undefined {
  const messages = ctx.messages || [];
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i].role === "user") return messages[i].content;
  }
  return undefined;
}

export default registerBeforeToolCallHook;

File: src/guardrail_hook.py
"""
Anti-Repeat-Errors Skill - Guardrail Hook Handler

Handles before_tool_call interception and rewriting.
"""

from __future__ import annotations

import json
import logging
import time
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Any, Optional

# Handle both package and direct imports
try:
    from .guardrail_models import GuardrailAction, GuardrailRule, GuardrailHit
    from .pattern_matcher import PatternMatcher, MatchResult
    from .logger import InjectionLogger, get_logger
except ImportError:
    from guardrail_models import GuardrailAction, GuardrailRule, GuardrailHit
    from pattern_matcher import PatternMatcher, MatchResult
    from logger import InjectionLogger, get_logger


@dataclass
class ToolCallContext:
    """工具调用上下文"""
    
    tool_name: str
    tool_params: dict[str, Any]
    session_key: Optional[str] = None
    phase: Optional[int] = None
    task_type: Optional[str] = None
    message_content: Optional[str] = None
    
    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class GuardrailResult:
    """Guardrail 处理结果"""
    
    # 核心结果
    allowed: bool  # 是否允许继续执行
    modified: bool  # 是否修改了参数
    
    # 工具信息
    tool_name: str
    original_params: dict[str, Any]
    result_params: dict[str, Any]
    
    # 规则信息
    rule_id: Optional[str] = None
    rule_name: Optional[str] = None
    action: Optional[str] = None  # rewrite/block/warn/log
    
    # 消息
    message: Optional[str] = None
    
    # 性能
    duration_ms: float = 0.0
    
    def to_dict(self) -> dict:
        return asdict(self)


class GuardrailHook:
    """
    Guardrail Hook 处理器
    
    负责:
    1. 在 before_tool_call 时进行模式匹配
    2. 执行拦截/改写动作
    3. 记录命中日志
    """
    
    def __init__(
        self,
        rules_dir: Optional[Path] = None,
        logger: Optional[InjectionLogger] = None,
    ):
        """
        初始化 Guardrail Hook
        
        Args:
            rules_dir: guardrail 规则目录
            logger: 日志记录器
        """
        self.matcher = PatternMatcher(rules_dir=rules_dir)
        self.logger = logger or get_logger()
        self._internal_logger = logging.getLogger("anti-repeat-errors.guardrail")
    
    def process_tool_call(self, context: ToolCallContext) -> GuardrailResult:
        """
        处理工具调用
        
        Args:
            context: 工具调用上下文
            
        Returns:
            GuardrailResult 对象
        """
        start = time.time()
        
        try:
            # 执行模式匹配
            match_result = self.matcher.match(
                tool_name=context.tool_name,
                tool_params=context.tool_params,
                context={
                    "session_key": context.session_key,
                    "phase": context.phase,
                    "task_type": context.task_type,
                    "message": context.message_content,
                }
            )
            
            duration_ms = (time.time() - start) * 1000
            
            # 无匹配，放行
            if not match_result.matched:
                return GuardrailResult(
                    allowed=True,
                    modified=False,
                    tool_name=context.tool_name,
                    original_params=context.tool_params,
                    result_params=context.tool_params,
                    duration_ms=duration_ms,
                )
            
            # 有匹配，根据动作处理
            rule = match_result.rule
            action = match_result.action
            
            # 记录命中日志
            self._log_hit(context, match_result)
            
            if action == GuardrailAction.BLOCK:
                # 阻断
                return GuardrailResult(
                    allowed=False,
                    modified=False,
                    tool_name=context.tool_name,
                    original_params=context.tool_params,
                    result_params=context.tool_params,
                    rule_id=rule.id,
                    rule_name=rule.name,
                    action=action.value,
                    message=match_result.message,
                    duration_ms=duration_ms,
                )
            
            elif action == GuardrailAction.REWRITE:
                # 改写
                return GuardrailResult(
                    allowed=True,
                    modified=True,
                    tool_name=context.tool_name,
                    original_params=context.tool_params,
                    result_params=match_result.result_params,
                    rule_id=rule.id,
                    rule_name=rule.name,
                    action=action.value,
                    message=match_result.message,
                    duration_ms=duration_ms,
                )
            
            elif action == GuardrailAction.WARN:
                # 警告但放行
                return GuardrailResult(
                    allowed=True,
                    modified=False,
                    tool_name=context.tool_name,
                    original_params=context.tool_params,
                    result_params=context.tool_params,
                    rule_id=rule.id,
                    rule_name=rule.name,
                    action=action.value,
                    message=match_result.message,
                    duration_ms=duration_ms,
                )
            
            else:  # LOG
                # 仅记录，放行
                return GuardrailResult(
                    allowed=True,
                    modified=False,
                    tool_name=context.tool_name,
                    original_params=context.tool_params,
                    result_params=context.tool_params,
                    rule_id=rule.id,
                    rule_name=rule.name,
                    action=action.value,
                    message=match_result.message,
                    duration_ms=duration_ms,
                )
                
        except Exception as e:
            self._internal_logger.error(f"Guardrail processing failed: {e}")
            # 错误时不阻断正常流程
            return GuardrailResult(
                allowed=True,
                modified=False,
                tool_name=context.tool_name,
                original_params=context.tool_params,
                result_params=context.tool_params,
                message=f"Guardrail error: {e}",
                duration_ms=(time.time() - start) * 1000,
            )
    
    def _log_hit(self, context: ToolCallContext, match_result: MatchResult) -> None:
        """记录命中日志"""
        from datetime import datetime
        
        hit = GuardrailHit(
            timestamp=datetime.now().isoformat(),
            rule_id=match_result.rule.id if match_result.rule else None,
            rule_name=match_result.rule.name if match_result.rule else None,
            action=match_result.action,
            tool_name=context.tool_name,
            original_params=context.tool_params,
            result_params=match_result.result_params,
            message=match_result.message,
            session_key=context.session_key,
            phase=context.phase,
        )
        
        # 写入命中日志文件
        self._write_hit_log(hit)
    
    def _write_hit_log(self, hit: GuardrailHit) -> None:
        """写入命中日志到文件"""
        log_dir = Path.home() / ".openclaw" / "logs" / "anti-repeat-errors"
        log_dir.mkdir(parents=True, exist_ok=True)
        
        log_file = log_dir / "guardrail_hits.jsonl"
        
        with open(log_file, "a", encoding="utf-8") as f:
            f.write(hit.model_dump_json())
            f.write("\n")


def create_guardrail_hook(
    rules_dir: Optional[Path] = None,
    logger: Optional[InjectionLogger] = None,
) -> GuardrailHook:
    """创建 Guardrail Hook 的便捷函数"""
    return GuardrailHook(rules_dir=rules_dir, logger=logger)